File: 2017/lua/day10.lua
local stuff = require("stuff")

local input = io.input("../__in/10.txt"):read(); io.close()

local p1data = stuff.str2numtable(input)
local p2data = {}
for _, v in ipairs(stuff.str2anychartable(input)) do table.insert(p2data, string.byte(v)) end
for _, v in ipairs({17, 31, 73, 47, 23}) do table.insert(p2data, v) end -- add standard ending

size = 256

-- part 1
local t = stuff.knothash(stuff.init(), p1data, 1)
print(string.format("Part 1: %d", t[1] * t[2]))

-- part 2
t = stuff.knothash(stuff.init(), p2data, 64)
print(string.format("Part 2: %s", stuff.densehash(t, hex)))

File: 2017/kotlin/day10.kt
import java.io.File

fun main() {
    val input = File("../__in/10.txt").readText().trim()
    val dataPart1 = input.split(",").mapNotNull { it.toIntOrNull() }
    val dataPart2 = input.map { it.code } + listOf(17, 31, 73, 47, 23)
    
    println("Part 1: ${with(knotHash(List(256) { i -> i }, dataPart1, 1)) { this[0] * this[1] }}")
    println("Part 2: ${denseHash(knotHash(List(256) { i -> i }, dataPart2, 64))}")
}